Infra: make assert_array_equals more useful

annevk opened the pull request:
Fixes part of #7179.

annevk committed:
Infra: make assert_array_equals more useful

Fixes part of #7179.

jgraham commented on resources/testharness.js:
```diff
@@ -1184,19 +1184,21 @@ policies and contribution forms [3].
                "assert_array_equals", description,
-               "lengths differ, expected ${expected} got ${actual}",
-               {expected:expected.length, actual:actual.length});
+               "lengths differ, expected array ${expected} got ${actual}",
```
The original reason for not doing this is concern about what happens when the array is large e.g. if you have something like `actual = []; actual[10000]=1`. So if we are going to print the array values, I'd like a helper function that checks if the array length is under some limit and if so prints the while thing, but otherwise just a limited number of elements close to an index (e.g. we might output full arrays up to 20 elements and then at most 20 elements in this case the last 20 elements, or something).

jgraham commented on resources/testharness.js:
```diff
@@ -1184,19 +1184,21 @@ policies and contribution forms [3].
             assert(actual.hasOwnProperty(i) === expected.hasOwnProperty(i),
                    "assert_array_equals", description,
-                   "property ${i}, property expected to be ${expected} but was ${actual}",
+                   "expected property ${i} to be ${expected} but was ${actual} (expected array ${arrayExpected} got ${arrayActual})",
```
Similar comments down here too about not printing too many elements.

annevk committed:
only stringify up to 21 values

annevk committed:
fix some issues, write tests, but wait with filling out expectations until there's agreement

annevk committed:
nit

jgraham committed:
Make assert_array_equals message maximise the size of array slice

When the output array is longer than the maximum size, try to choose a
slice centred on the offset with the maximum possible size. For
example if the array has N_MAX + 10 elements and the offset is 2 from
the end, we choose the final 2 elements, plus the offset plus the
N_MAX - 3 elements before the offset. If the offset is in the middle
of the same array, we take N_MAX/2 elements before and after the
offset.

jgraham committed:
Add expected data for assert_array_equals test

jgraham commented:
The infra changes will land in a separate PR, but I've added them here since the variant changes are required, and the rest can be conveniently tested. Once it's clear they work in this context I'll do the separation.

annevk commented on resources/testharness.js:
```diff
@@ -1178,25 +1186,58 @@ policies and contribution forms [3].
-               {expected:expected.length, actual:actual.length});
+               "lengths differ, expected array ${expected} length ${expectedLength}, got ${actual} length ${actualLength}",
+               {expected:shorten_array(expected, expected.length - 1),
+                expectedLength:expected.length,
```
You use different indentation rules here than elsewhere. (I like this style better, but...)

jgraham replied:
I honestly have no idea what the rules for this file are; I don't think we should block on this (but should look at making the formatting better at some point).

annevk commented on resources/testharness.js:
```diff
@@ -1178,25 +1186,58 @@ policies and contribution forms [3].
+            // if it takes us before the start we have more after.
+            const length_after_offset = Math.floor(max_array_length / 2);
+            let upper_bound = Math.min(length_after_offset + offset, arr.length);
+            let lower_bound = Math.max(upper_bound - max_array_length, 0);
```
This could be const.

annevk reviewed (commented):
I'm happy with this, but can't approve as it's my own PR.
